synapse/storage/data_stores/main/events_worker.py

```diff
@@ -1358,6 +1363,70 @@ def get_next_event_to_expire_txn(txn):
             desc="get_next_event_to_expire", func=get_next_event_to_expire_txn
         )
 
+    @cached(tree=True, cache_context=True)
+    async def get_unread_message_count_for_user(
+        self, room_id: str, user_id: str, cache_context: _CacheContext,
+    ):
+        last_read_event_id = await self.get_last_receipt_event_id_for_user(
+            user_id=user_id,
+            room_id=room_id,
+            receipt_type="m.read",
+            on_invalidate=cache_context.invalidate,
+        )
+
+        return await self.db.runInteraction(
+            "get_unread_message_count_for_user",
+            self._get_unread_message_count_for_user_txn,
+            user_id,
+            room_id,
+            last_read_event_id,
+        )
+
+    def _get_unread_message_count_for_user_txn(
+        self,
+        txn: LoggingTransaction,
+        user_id: str,
+        room_id: str,
+        last_read_event_id: Optional[str],
+    ):
+        if last_read_event_id:
+            # Get the stream ordering for the last read event.
+            stream_ordering = self.db.simple_select_one_onecol_txn(
+                txn=txn,
+                table="events",
+                keyvalues={"room_id": room_id, "event_id": last_read_event_id},
+                retcol="stream_ordering",
+            )
+        else:
+            # If there's no read receipt for that room, it probably means the user hasn't
+            # opened it yet, in which case use the stream ID of their join event.
+            # We can't just set it to 0 otherwise messages from other local users from
+            # before this user joined will be counted as well.
+            txn.execute(
+                """
+                SELECT stream_ordering FROM local_current_membership
+                LEFT JOIN events USING (event_id, room_id)
+                WHERE membership = 'join'
+                    AND user_id = ?
+                    AND room_id = ?
+                """,
+                (user_id, room_id),
+            )
+            row = txn.fetchone()
+            stream_ordering = row[0]
```

Comment:
looks like this will explode horribly if the user turns out not to be a current member of the room. suggest failing gracefully somehow, or if you're really sure it shouldn't happen adding an assertion so that when it does the error message is better than obscure "ValueError".